### ISSUE TITLE
Edit macro

### ISSUE THREAD

**Issue by liketoeatcheese** [svermeulen/vim-macrobatics#13]
Hey mate,

Sweet macro plugin :) It's been a joy using it. I'm wondering if there's a way to edit the macro directly?

**Comment by svermeulen** [comment]
Thanks :)

It still saves macros to registers, so you can edit the register directly (whichever register you are using for `g:Mac_DefaultRegister`).  And if saving macros persistently, you can edit the generated file.  Personally I do not though, because it's common in my case for my macros to contain a bunch of non printable characters.  Hoping that eventually neovim improves this as discussed [here](https://github.com/neovim/neovim/issues/14732).  What I do instead is either add logic by either prepending or appending, or if neither of those are possible then re-record the whole thing

**Comment by liketoeatcheese** [comment]
Since I asked the question, I have been digging into your code base a little. I noticed that you use `writefile([a:microdata],a:filename,'b')` which write file from register to binary file, and `readfile(a:filename','b')`. Is it worthwhile to create something like `<plug>(Mac_EditNamedMacro)`, which `fzf` to the NamedMacro (file based or global), if the user chooses it:
1. `let a = readfile(a:NamedMacroFileName','b')
2. call setreg('g:Mac_DefaultRegister', a[0])

Which will copy the named macro to the current default register. The user can then edit the register directly with something like `:let @defaultregister = <C-R><C-R>{default register}`. 

And yes, editing the binary directly is indeed funky. I tried to edit the binary file directly, and the weird stuff is when I tried to echo the content of the named macro as binary file again with readfile, instead of the original list:
 `['content of named register']`
 I got `['new content of named register','']`

**Comment by svermeulen** [comment]
Sounds like `<plug>(Mac_SearchForNamedMacroAndSelect)` might be what you're looking for?

Yeah, I had to use binary files to ensure that it played exactly the same as the original recording.  I expect that only some macros will be successfully editable

**Comment by liketoeatcheese** [comment]
Sweet!!! That works!

I tested something just now, and it may be helpful to you (I haven't attempted this with parameterized macro as I haven't tried them). Most of my macros are also long, but I find it quite helpful to edit them instead of scratching from the start. Say you want to edit a macro, use `q:` to enter vim command line, where you can somewhat move around as a buffer (Make it easier to move around and edit). Leverage `let @register_name = ""`. In edit mode, instead of `<C-R><C-R>{register_name}`,  you can use `<C-R><C-O>{register_name}` instead. Which will keep the content of the register as is, [a bug fix](https://stackoverflow.com/questions/20617329/in-vim-what-is-the-difference-between-c-rregister-and-c-rc-pregister-wi#:~:text=CTRL%2DR%20CTRL%2DO%20and%20the%20.%20command) introduced by Bram himself (RIP). A common case that I have seen is when I do search and replace, I use `\` a lot, like `\w` for regex. You have to escape these, so you can search and replace them to escape all since the command line is now treated as a buffer (Something like `'<,'>s/\\/\\\\/g`). And for places which you used `<Ctrl-{a character}>` (e.g `<Ctrl-j>` which vim doesn't fully capture them with `<Ctrl-R><Ctrl-O>`, you can simply `<Ctrl-v>` and replace that motion instead (e.g `<Ctrl-v><Ctrl-j>`)

Since you are here, is there an option to turn off parameterized? (One option to save without the "Add parameters?" prompt, and one option to save with it)?

**Comment by svermeulen** [comment]
Interesting that editing in the command line window allows editing macros more reliably.  I'll have to try that!

Most of my macros are parameterized, so I haven't felt the need for that, but happy to make this an option.  We could add a flag to the methods here:

```
nnoremap <plug>(Mac_NameCurrentMacro) :<c-u>call macrobatics#nameCurrentMacro()<cr>
nnoremap <plug>(Mac_NameCurrentMacroForFileType) :<c-u>call macrobatics#nameCurrentMacroForFileType()<cr>
nnoremap <plug>(Mac_NameCurrentMacroForCurrentSession) :<c-u>call macrobatics#nameCurrentMacroForCurrentSession()<cr>
```

And forward it to `s:saveCurrentMacroToDirectory`

**Comment by liketoeatcheese** [comment]
That would be perfect! Since you are more familiar with the code base than I am

**Comment by svermeulen** [comment]
Ok will take a look.  My vimscript has become pretty rusty since I wrote this but should be straightforward

**Comment by svermeulen** [comment]
Can you try branch `sverm/skip-parameter-prompt` and change your config to something like this:

```
nnoremap <plug>(Mac_NameCurrentMacro) :<c-u>call macrobatics#nameCurrentMacro(1)<cr>
nnoremap <plug>(Mac_NameCurrentMacroForFileType) :<c-u>call macrobatics#nameCurrentMacroForFileType(1)<cr>
nnoremap <plug>(Mac_NameCurrentMacroForCurrentSession) :<c-u>call macrobatics#nameCurrentMacroForCurrentSession(1)<cr>
```

**Comment by liketoeatcheese** [comment]
![image](https://github.com/svermeulen/vim-macrobatics/assets/88610355/29b10053-4a9a-4432-a036-0f534e6e9b49)

I'm getting this. Here's my config:
```
local opts = { noremap = true, silent = true }
keymap("n", "<plug>(Mac_NameCurrentMacro)", ":<c-u>call macrobatics#nameCurrentMacro(1)<cr>", opts)
keymap("n", "<plug>(Mac_NameCurrentMacroForFileType)", ":<c-u>call macrobatics#nameCurrentMacroForFileType(1)<cr>", opts)
keymap("n", "<plug>(Mac_NameCurrentMacroForCurrentSession)", ":<c-u>call macrobatics#nameCurrentMacroForCurrentSession(1)<cr>", opts)

```

**Comment by svermeulen** [comment]
Latest commit on branch `sverm/skip-parameter-prompt`?

**Comment by liketoeatcheese** [comment]
Yeb:
	use({ "svermeulen/vim-macrobatics", branch = "sverm/skip-parameter-prompt" })
	This will basically pull the latest commit

I only see one commit. Maybe you have other ones, did you push?

**Comment by liketoeatcheese** [comment]
Ah wait, I use vim.cmd([[]]) instead and it worked!

**Comment by liketoeatcheese** [comment]
By the way, I found a way to edit without doing any transformation while keeping all special characters as is! I implemented it in Neovim with lua, but if you are using vim and good with vimscript, I'm sure you can convert it somehow:
```
-- Define a function to edit a macro from a register in a horizontal split
_G.EditMacroFromRegister = function(register)
  print("Macro from register: "..register)
  -- Get the content of the specified register
  local macro_content = vim.fn.getreg(register)

  vim.ui.input({
    prompt = "Edit macro: ",
    default = macro_content,
    completion = "file",
    highlight = function(input)
      if string.len(input) > 8 then
        return { { 0, 8, "InputHighlight" } }
      else
        return {}
      end
    end,
  }, function(input)
    if input then
      print("Saved to register with content: " .. input)
        vim.fn.setreg(register, input)
    else
      print "You cancelled"
    end
  end)
end

-- map our provider to a user command ':Directories'
vim.cmd([[command! -nargs=* EditMacroFromRegister lua _G.EditMacroFromRegister("m")]])

-- or to a keybind, both below are (sort of) equal
vim.keymap.set('n', '<Leader>me', ":EditMacroFromRegister<cr>", {desc = "Edit default macro"})
```

This is prompted as input; whatever you enter (save) will be returned to the register as is! You can switch to normal mode and move around like a buffer. In this case, I have my default macro register to "m"

**Comment by svermeulen** [comment]
I should clarify that when I suggested this:

```
nnoremap <plug>(Mac_NameCurrentMacro) :<c-u>call macrobatics#nameCurrentMacro(1)<cr>
nnoremap <plug>(Mac_NameCurrentMacroForFileType) :<c-u>call macrobatics#nameCurrentMacroForFileType(1)<cr>
nnoremap <plug>(Mac_NameCurrentMacroForCurrentSession) :<c-u>call macrobatics#nameCurrentMacroForCurrentSession(1)<cr>
```

This was incorrect. What I meant was this:

```
nnoremap <leader>mn :<c-u>call macrobatics#nameCurrentMacro(1)<cr>
nnoremap <leader>mf :<c-u>call macrobatics#nameCurrentMacroForFileType(1)<cr>
nnoremap <leader>ms :<c-u>call macrobatics#nameCurrentMacroForCurrentSession(1)<cr>
```

Using whatever keys you want in place of the leader keys

I think you'll get errors if you remap the plugs themselves

Will have to try editing directly again like you say, sounds really useful

**Comment by liketoeatcheese** [comment]
You were right. I set up as you said and having this after doing `macrobatics#nameCurrentMacroForFileType(1)`:
![image](https://github.com/svermeulen/vim-macrobatics/assets/88610355/13ba9e2e-619f-4a9b-bb5f-a8d658dbd817)

Here's my set up:
```lua
use({ "svermeulen/vim-macrobatics", branch = "sverm/skip-parameter-prompt" })
```
```lua
vim.cmd([[
nnoremap <leader>mn :<c-u>call macrobatics#nameCurrentMacroForFileType(1)<cr>
nmap <nowait> mq <plug>(Mac_RecordNew)
]])
```

**Comment by liketoeatcheese** [comment]
Yea, that works! Thank you!

Do you reckon if adding the functionality I mentioned above is viable, can we have another function to edit it directly? Like `function! macrobatics#editmacro()<cr>` (Not sure if that will affect the parameterized one). I have it working on my side but it's more like multi-step process: `<plug>(Mac_SearchForNamedMacroAndSelect)` -> `EditMacroFromRegister` -> `<plug>(Mac_SearchForNamedMacroAndOverwrite)`

**Comment by svermeulen** [comment]
Yeah I agree, a way to edit in a scratch buffer would be great feature

**Comment by svermeulen** [comment]
Btw what plugin are you using that provides those error popup windows?

**Comment by liketoeatcheese** [comment]
Thanks for that :)

I use https://github.com/folke/noice.nvim. They have a pretty nice setup but I'm not keen on the pop-up command line so I set it back to the bottom, you will notice when you set it up. Here's my set up:
```lua
local noice_status_ok, noice = pcall(require, "noice")
if not noice_status_ok then
	return
end

noice.setup({
	cmdline = {
		view = "cmdline",
	},
	lsp = {
		-- override markdown rendering so that **cmp** and other plugins use **Treesitter**
		override = {
			["vim.lsp.util.convert_input_to_markdown_lines"] = true,
			["vim.lsp.util.stylize_markdown"] = true,
			["cmp.entry.get_documentation"] = true,
		},
	},
	routes = { { view = "cmdline", filter = { event = "msg_showmode" } } },
	-- you can enable a preset for easier configuration
	presets = {
		bottom_search = true, -- use a classic bottom cmdline for search
		command_palette = true, -- position the cmdline and popupmenu together
		long_message_to_split = true, -- long messages will be sent to a split
		inc_rename = false, -- enables an input dialog for inc-rename.nvim
		lsp_doc_border = false, -- add a border to hover docs and signature help
	},
})
```

**Comment by liketoeatcheese** [comment]
Hey @svermeulen, just found a bug. If I edit the register "m" directly which is the main one, and save it to the NamedRegister, when I play the register, I got this (Note that playing the macro in the register with `@m` still works as expected):
![image](https://github.com/svermeulen/vim-macrobatics/assets/88610355/133958f4-21f0-4dd1-a637-bd989dc503cb)

I checked the code and this line was there:
```vim
    " Don't need to use i here though, because we only want this to run at the very end
    " Since otherwise it will overwrite s:repeatMacro
    call feedkeys("\<plug>(Mac__OnPlayMacroCompleted)", 'm')
```

I commented it out in the plugin local repo .local/share... directory and the error disappeared. Any thoughts?
